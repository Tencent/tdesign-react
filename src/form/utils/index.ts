import type { NamePath } from '../type';

// 获取 formMap 管理的数据
export function getMapValue(name: NamePath, formMapRef: React.MutableRefObject<Map<any, any>>) {
  if (!formMapRef.current) return;

  // 提取所有 map key
  const mapKeys = [...formMapRef.current.keys()];
  // 转译为字符串后比对 key 兼容数组格式
  const key = mapKeys.find((key) => String(key) === String(name));
  // 拿到 key 引用地址获取 value
  return formMapRef.current.get(key);
}

// 将数据整理成对象，数组 name 转化嵌套对象: ['user', 'name'] => { user: { name: '' } }
export function calcFieldValue(name: NamePath, value: any) {
  if (Array.isArray(name)) {
    const fieldValue = name.reduceRight((prev, curr) => {
      const arr = [];
      if (/^\d+$/.test(String(curr))) arr[curr] = prev;
      return arr.length ? arr : { [curr]: prev };
    }, value);
    return { ...fieldValue };
  }

  return { [name]: value };
}

// 通过对象数据类型获取 map 引用: { user: { name: '' } } => formMap.get(['user', 'name'])
export function travelMapFromObject(
  obj: Record<any, any>,
  formMapRef: React.MutableRefObject<Map<any, any>>,
  callback: Function,
) {
  for (const [mapName, formItemRef] of formMapRef.current.entries()) {
    // 支持嵌套数据结构
    if (Array.isArray(mapName)) {
      // 创建唯一临时变量 symbol
      const symbol = Symbol('name');
      let fieldValue = null;

      for (let i = 0; i < mapName.length; i++) {
        const item = mapName[i];
        if (Reflect.has(fieldValue || obj, item)) {
          fieldValue = Reflect.get(fieldValue || obj, item);
        } else {
          // 当反射无法获取到值则重置为 symbol
          fieldValue = symbol;
          break;
        }
      }

      // 非 symbol 说明获取到了值
      if (fieldValue !== symbol) {
        callback(formItemRef, fieldValue);
      }
    } else if (Reflect.has(obj, mapName)) {
      callback(formItemRef, obj[mapName]);
    }
  }
}
