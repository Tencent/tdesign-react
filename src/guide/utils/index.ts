import { scrollToParentVisibleArea } from './getScrollParent';
import getRelativePosition from './getRelativePosition';
import getTargetElm from './getTargetElm';
import scrollToElm from './scrollToElm';
import useMount from './useMount';

export { scrollToParentVisibleArea, getRelativePosition, getTargetElm, scrollToElm, useMount };
