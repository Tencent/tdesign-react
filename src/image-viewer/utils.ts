export const downloadFile = function (imgSrc) {
  const image = new Image();
  const name = imgSrc?.split?.('/').pop() || Math.random().toString(32).slice(2);

  image.setAttribute('crossOrigin', 'anonymous');

  image.onload = () => {
    const canvas = document.createElement('canvas');
    canvas.width = image.width;
    canvas.height = image.height;

    const context = canvas.getContext('2d');
    context.drawImage(image, 0, 0, image.width, image.height);
    canvas.toBlob((blob) => {
      const url = URL.createObjectURL(blob);
      const a = document.createElement('a');
      a.download = name;
      a.href = url;
      a.click();
      a.remove();
      URL.revokeObjectURL(url);
    });
  };
  image.src = imgSrc;
};
