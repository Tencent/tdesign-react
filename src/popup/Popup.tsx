import React, {
  forwardRef,
  CSSProperties,
  useState,
  cloneElement,
  isValidElement,
  useMemo,
  useCallback,
  useRef,
  useEffect,
} from 'react';
import { CSSTransition } from 'react-transition-group';
import classNames from 'classnames';
import { usePopper } from 'react-popper';
import { Placement } from '@popperjs/core';
import { StyledProps } from '../common';
import useControlled from '../hooks/useControlled';
import useAnimation from '../_util/useAnimation';
import useConfig from '../_util/useConfig';
import composeRefs from '../_util/composeRefs';
import { TdPopupProps } from './type';
import Portal from '../common/Portal';
import useTriggerProps from './hooks/useTriggerProps';
import getTransitionParams from './utils/getTransitionParams';
import useMutationObserver from '../_util/useMutationObserver';
import useWindowSize from '../_util/useWindowSize';
import { popupDefaultProps } from './defaultProps';

export interface PopupProps extends TdPopupProps, StyledProps {
  // 是否触发展开收起动画，内部下拉式组件使用
  expandAnimation?: boolean;
  // 初始化popper的可定制option
  popperModifiers?: Array<{ name: string; options: Object }>;
  updateScrollTop?: (content: HTMLDivElement) => void;
}

function getPopperPlacement(placement: TdPopupProps['placement']) {
  return placement.replace(/-(left|top)$/, '-start').replace(/-(right|bottom)$/, '-end') as Placement;
}

const Popup = forwardRef((props: PopupProps, ref: React.Ref<HTMLDivElement>) => {
  const {
    trigger,
    content,
    placement,
    attach,
    showArrow,
    destroyOnClose,
    className,
    style,
    overlayClassName,
    overlayStyle,
    triggerElement,
    children = triggerElement,
    disabled,
    zIndex,
    onVisibleChange,
    onScroll,
    expandAnimation,
    popperModifiers = [],
    updateScrollTop,
  } = props;
  const { classPrefix } = useConfig();

  // 全局配置
  const { keepExpand, keepFade } = useAnimation();

  const [visible, setVisible] = useControlled(props, 'visible', onVisibleChange);

  const { height: windowHeight, width: windowWidth } = useWindowSize();

  // refs
  const [triggerRef, setTriggerRef] = useState<HTMLElement>(null);
  const [overlayRef, setOverlayRef] = useState<HTMLDivElement>(null);

  const contentRef = useRef<HTMLDivElement>(null);
  const referenceRef = useRef<HTMLDivElement>(null);
  const popupRef = useRef(null);

  const popperRef = useRef(null);
  const portalRef = useRef(null);

  // 展开时候动态判断上下左右翻转
  const onPopperFirstUpdate = useCallback((state) => {
    const referenceElmRect = referenceRef.current.getBoundingClientRect();
    const { top: referenceElmTop, left: referenceElmLeft, bottom, right } = referenceElmRect;
    const referenceElmBottom = window.innerHeight - bottom;
    const referenceElmRight = window.innerWidth - right;

    const { offsetHeight: contentOffsetHeight, offsetWidth: contentOffsetWidth } = contentRef.current;

    let newPlacement = state.options.placement;
    // 底部不够向上翻转
    if (referenceElmBottom < contentOffsetHeight && referenceElmTop >= contentOffsetHeight) {
      newPlacement = state.options.placement.replace('bottom', 'top');
    }
    // 顶部不够向下翻转
    if (referenceElmTop < contentOffsetHeight && referenceElmBottom >= contentOffsetHeight) {
      newPlacement = state.options.placement.replace('top', 'bottom');
    }
    // 左侧不够向右翻转
    if (referenceElmLeft < contentOffsetWidth && referenceElmRight >= contentOffsetWidth) {
      newPlacement = state.options.placement.replace('left', 'right');
    }
    // 右侧不够向左翻转
    if (referenceElmRight < contentOffsetWidth && referenceElmLeft >= contentOffsetWidth) {
      newPlacement = state.options.placement.replace('right', 'left');
    }
    Object.assign(state.options, { ...state.options, placement: newPlacement });
    popperRef.current.update();
  }, []);

  const options = useMemo(
    () => ({
      placement: getPopperPlacement(placement),
      onFirstUpdate: onPopperFirstUpdate,
      modifiers: popperModifiers,
    }),
    [onPopperFirstUpdate, placement, popperModifiers],
  );

  popperRef.current = usePopper(triggerRef, overlayRef, options);

  const { styles, attributes } = popperRef.current;
  const defaultStyles = useMemo(() => {
    if (triggerRef && typeof overlayStyle === 'function') return { ...overlayStyle(triggerRef, overlayRef) };
    return { ...overlayStyle };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [overlayStyle, triggerRef, overlayRef, visible]);

  // 设置 style 决定展示与隐藏
  const overlayVisibleStyle: CSSProperties = defaultStyles;

  const triggerNodeTemp = useMemo(() => {
    const [triggerChildNode] = React.Children.toArray(children);

    if (React.Children.count(children) === 1 && isValidElement(triggerChildNode)) {
      return triggerChildNode;
    }
    return <span className={`${classPrefix}-trigger`}>{children}</span>;
  }, [children, classPrefix]);

  const [triggerProps, popupProps] = useTriggerProps(
    { current: overlayRef },
    { current: triggerRef },
    [trigger],
    visible,
    setVisible,
    disabled,
    triggerNodeTemp,
  );

  // 代理 trigger 的 ref
  const triggerNode = cloneElement(triggerNodeTemp, {
    ref: composeRefs((triggerNodeTemp as any).ref, setTriggerRef),
    ...triggerProps,
  });

  const handleExited = () => {
    if (!destroyOnClose) {
      portalRef.current.style.display = 'none';
    }
  };
  const handleEnter = () => {
    if (!destroyOnClose) {
      portalRef.current.style.display = 'block';
    }
  };

  // 监听 trigger 节点或内容变化动态更新 popup 定位
  useMutationObserver(triggerRef, () => {
    popperRef.current.update?.();
  });
  useEffect(() => {
    popperRef.current.update?.();
  }, [content, visible, windowHeight, windowWidth]);

  useEffect(() => {
    if (visible && overlayRef) {
      updateScrollTop?.(contentRef?.current);
    }
  }, [visible, overlayRef, updateScrollTop]);

  // 初次不渲染.
  const portal =
    visible || overlayRef ? (
      <CSSTransition
        appear
        in={visible}
        timeout={180}
        nodeRef={popupRef}
        unmountOnExit={destroyOnClose}
        onEnter={handleEnter}
        onExited={handleExited}
      >
        <Portal triggerNode={triggerRef} attach={attach} ref={portalRef}>
          <CSSTransition
            appear
            timeout={0}
            in={visible}
            nodeRef={popupRef}
            {...getTransitionParams({
              classPrefix,
              expandAnimation: expandAnimation && keepExpand,
              fadeAnimation: keepFade,
            })}
          >
            <div
              ref={composeRefs(setOverlayRef, ref, popupRef)}
              style={{ ...styles.popper, zIndex }}
              className={`${classPrefix}-popup`}
              {...attributes.popper}
              {...popupProps}
            >
              <div
                className={classNames(
                  `${classPrefix}-popup__content`,
                  {
                    [`${classPrefix}-popup__content--arrow`]: showArrow,
                  },
                  overlayClassName,
                )}
                style={overlayVisibleStyle}
                ref={contentRef}
                onScroll={(e) => onScroll?.({ e: e as React.WheelEvent<HTMLDivElement> })}
              >
                {showArrow ? <div style={styles.arrow} className={`${classPrefix}-popup__arrow`} /> : null}
                {content}
              </div>
            </div>
          </CSSTransition>
        </Portal>
      </CSSTransition>
    ) : null;

  return (
    <div className={classNames(`${classPrefix}-popup__reference`, className)} style={style} ref={referenceRef}>
      {triggerNode}
      {portal}
    </div>
  );
});

Popup.displayName = 'Popup';
Popup.defaultProps = popupDefaultProps;

export default Popup;
