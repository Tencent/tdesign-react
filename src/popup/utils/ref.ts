import { isMemo } from 'react-is';

// 合并 refs
export function composeRef<T>(...refs: React.Ref<T>[]): React.Ref<T> {
  return (node: T) => {
    refs.forEach((ref) => {
      if (typeof ref === 'function') {
        ref(node);
      } else if (typeof ref === 'object' && ref && 'current' in ref) {
        if (node && 'currentElement' in node) {
          // eslint-disable-next-line
          (ref as any).current = node['currentElement'];
        } else {
          // eslint-disable-next-line
          (ref as any).current = node;
        }
      }
    });
  };
}

// 判断是否支持 ref 透传
export function supportRef(nodeOrComponent: any): boolean {
  const type = isMemo(nodeOrComponent) ? nodeOrComponent.type.type : nodeOrComponent.type;

  // Function component node
  if (typeof type === 'function' && !type.prototype?.render) {
    return false;
  }

  // Class component
  if (typeof nodeOrComponent === 'function' && !nodeOrComponent.prototype?.render) {
    return false;
  }

  return true;
}
