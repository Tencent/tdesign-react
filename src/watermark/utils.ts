const toLowercaseSeparator = (key: string) => key.replace(/([A-Z])/g, '-$1').toLowerCase();

// style对象转字符串
export const getStyleStr = (style: React.CSSProperties): string =>
  Object.keys(style)
    .map((key: keyof React.CSSProperties) => `${toLowercaseSeparator(key)}: ${style[key]};`)
    .join(' ');
