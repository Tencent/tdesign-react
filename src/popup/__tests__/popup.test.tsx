import React, { useState } from 'react';
import { testExamples, render, act, fireEvent, waitFor } from '@test/utils';
import Popup from '../Popup';

// 测试组件代码 Example 快照
testExamples(__dirname);

describe('Popup 组件测试', () => {
  const popupText = '弹出层内容';
  const popupTestId = 'popup-test-id';
  const triggerElement = '触发元素';

  test('hover 触发测试', async () => {
    const { getByText, queryByTestId } = render(
      <Popup placement="top" content={<div data-testid={popupTestId}>{popupText}</div>}>
        {triggerElement}
      </Popup>,
    );

    // 鼠标进入前，没有元素存在
    const popupElement1 = await waitFor(() => queryByTestId(popupTestId));
    expect(popupElement1).toBeNull();

    // 模拟鼠标进入
    act(() => {
      fireEvent.mouseEnter(getByText(triggerElement));
      jest.runAllTimers();
    });

    // 鼠标进入后，有元素，而且内容为 popupText
    const popupElement2 = await waitFor(() => queryByTestId(popupTestId));

    expect(popupElement2).not.toBeNull();
    expect(popupElement2).toHaveTextContent(popupText);
    expect(popupElement2.parentNode.parentNode).toHaveClass('t-popup--animation-enter-active');

    // 模拟鼠标离开
    act(() => {
      fireEvent.mouseLeave(getByText(triggerElement));
      jest.runAllTimers();
    });

    // 鼠标离开，style 的 display 应该为 none
    const popupElement3 = await waitFor(() => queryByTestId(popupTestId));
    expect(popupElement3).not.toBeNull();
    expect(popupElement3.parentNode.parentNode).toHaveClass('t-popup--animation-leave-active');
  });

  test('click 触发测试', async () => {
    const { getByText, queryByTestId } = render(
      <Popup trigger="click" placement="top" content={<div data-testid={popupTestId}>{popupText}</div>}>
        {triggerElement}
      </Popup>,
    );

    // 鼠标进入前，没有元素存在
    const popupElement1 = await waitFor(() => queryByTestId(popupTestId));
    expect(popupElement1).toBeNull();

    // 模拟鼠标点击
    act(() => {
      fireEvent.click(getByText(triggerElement));
      jest.runAllTimers();
    });

    // 鼠标进入后，有元素，而且内容为 popupText
    const popupElement2 = await waitFor(() => queryByTestId(popupTestId));
    expect(popupElement2).not.toBeNull();
    expect(popupElement2).toHaveTextContent(popupText);
    expect(popupElement2.parentNode.parentNode).toHaveClass('t-popup--animation-enter-active');

    // 点击浮层
    act(() => {
      fireEvent.mouseDown(queryByTestId(popupTestId));
      jest.runAllTimers();
    });

    // 点击浮层也不隐藏
    const popupElement3 = await waitFor(() => queryByTestId(popupTestId));
    expect(popupElement3.parentNode.parentNode).toHaveClass('t-popup--animation-enter-active');

    // 模拟鼠标点击其他地方
    act(() => {
      fireEvent.mouseDown(document);
      jest.runAllTimers();
    });

    // 鼠标离开，style 的 display 应该为 none
    const popupElement4 = await waitFor(() => queryByTestId(popupTestId));
    expect(popupElement4).not.toBeNull();
    expect(popupElement4.parentNode.parentNode).toHaveClass('t-popup--animation-leave-active');
  });

  test('focus 触发测试', async () => {
    const { getByText, queryByTestId } = render(
      <Popup trigger="focus" placement="top" content={<div data-testid={popupTestId}>{popupText}</div>}>
        {triggerElement}
      </Popup>,
    );

    // 鼠标进入前，没有元素存在
    const popupElement1 = await waitFor(() => queryByTestId(popupTestId));
    expect(popupElement1).toBeNull();

    // 模拟鼠标聚焦
    act(() => {
      fireEvent.focus(getByText(triggerElement));
      jest.runAllTimers();
    });

    // 鼠标进入后，有元素，而且内容为 popupText
    const popupElement2 = await waitFor(() => queryByTestId(popupTestId));
    expect(popupElement2).not.toBeNull();
    expect(popupElement2).toHaveTextContent(popupText);
    expect(popupElement2.parentNode.parentNode).toHaveClass('t-popup--animation-enter-active');

    // 模拟鼠标失焦
    act(() => {
      fireEvent.blur(getByText(triggerElement));
      jest.runAllTimers();
    });

    // 鼠标离开，style 的 display 应该为 none
    const popupElement3 = await waitFor(() => queryByTestId(popupTestId));
    expect(popupElement3).not.toBeNull();
    expect(popupElement3.parentNode.parentNode).toHaveClass('t-popup--animation-leave-active');
  });

  test('contextMenu 触发测试', async () => {
    const { getByText, queryByTestId } = render(
      <Popup trigger="context-menu" placement="top" content={<div data-testid={popupTestId}>{popupText}</div>}>
        {triggerElement}
      </Popup>,
    );

    // 鼠标进入前，没有元素存在
    const popupElement1 = await waitFor(() => queryByTestId(popupTestId));
    expect(popupElement1).toBeNull();

    // 模拟鼠标右键
    act(() => {
      fireEvent.contextMenu(getByText(triggerElement));
      jest.runAllTimers();
    });

    // 鼠标进入后，有元素，而且内容为 popupText
    const popupElement2 = await waitFor(() => queryByTestId(popupTestId));
    expect(popupElement2).not.toBeNull();
    expect(popupElement2).toHaveTextContent(popupText);
    expect(popupElement2.parentNode.parentNode).toHaveClass('t-popup--animation-enter-active');

    // 模拟鼠标点击其他地方
    act(() => {
      fireEvent.mouseDown(document);
      jest.runAllTimers();
    });

    // 鼠标离开，style 的 display 应该为 none
    const popupElement3 = await waitFor(() => queryByTestId(popupTestId));
    expect(popupElement3).not.toBeNull();
    expect(popupElement3.parentNode.parentNode).toHaveClass('t-popup--animation-leave-active');
  });

  test('测试隐藏后销毁', async () => {
    const { getByText, queryByTestId } = render(
      <Popup destroyOnClose placement="top" content={<div data-testid={popupTestId}>{popupText}</div>}>
        {triggerElement}
      </Popup>,
    );

    // 模拟鼠标进入
    act(() => {
      fireEvent.mouseEnter(getByText(triggerElement));
      jest.runAllTimers();
    });

    // 鼠标进入后，有元素，而且内容为 popupText
    const popupElement2 = await waitFor(() => queryByTestId(popupTestId));
    expect(popupElement2).not.toBeNull();

    // 模拟鼠标离开
    act(() => {
      fireEvent.mouseLeave(getByText(triggerElement));
      jest.runAllTimers();
    });

    // 鼠标离开，style 的 display 应该为 none
    const popupElement3 = await waitFor(() => queryByTestId(popupTestId));
    expect(popupElement3).toBeNull();
  });

  test('渲染到指定位置', async () => {
    // 提前创建一个容器
    const specialId = 'SPECIAL_ID';
    const $container = document.createElement('div');
    $container.id = specialId;
    document.body.appendChild($container);

    const { queryByTestId } = render(
      <Popup
        visible
        placement="top"
        attach={() => $container}
        content={<div data-testid={popupTestId}>{popupText}</div>}
      >
        {triggerElement}
      </Popup>,
    );

    // 弹出层的父元素的 id 为 SPECIAL_ID
    const popupElement = await waitFor(() => queryByTestId(popupTestId));
    expect(popupElement).not.toBeNull();

    expect(
      // 注意，这里层级有点多
      // #SPECIAL_ID
      // └-> .div
      //   └-> .t-popup
      //     └-> div popper 定位层
      //       └-> div[data-testid="popup-test-id"]
      popupElement.parentElement.parentElement.parentElement.parentElement,
    ).toHaveAttribute('id', specialId);
  });

  test('受控展示&隐藏', async () => {
    const testShowButton = 'test-show-button';
    const testHideButton = 'test-hide-button';

    function TestComponent() {
      const [visible, setVisible] = useState(false);
      const $content = (
        <button data-testid={testHideButton} onClick={() => setVisible(false)}>
          隐藏
        </button>
      );

      return (
        <>
          <Popup destroyOnClose visible={visible} content={$content}>
            <button data-testid={testShowButton} onClick={() => setVisible(true)}>
              显示
            </button>
          </Popup>
        </>
      );
    }

    const { queryByTestId } = render(<TestComponent />);

    // 鼠标点击前，没有元素存在
    const popupElement1 = await waitFor(() => queryByTestId(testHideButton));
    expect(popupElement1).toBeNull();

    // 模拟鼠标点击显示按钮
    act(() => {
      fireEvent.click(queryByTestId(testShowButton));
      jest.runAllTimers();
    });

    // 隐藏的按钮展示出来
    const popupElement2 = await waitFor(() => queryByTestId(testHideButton));
    expect(popupElement2).not.toBeNull();

    // 模拟鼠标点击隐藏按钮
    act(() => {
      fireEvent.click(queryByTestId(testHideButton));
      jest.runAllTimers();
    });

    // 隐藏的按钮消失
    const popupElement3 = await waitFor(() => queryByTestId(testHideButton));
    expect(popupElement3).toBeNull();
  });

  test('测试浮层附加属性及 content 为空', async () => {
    const testClassName = 'test-class-name';
    const testStyle = { color: '#ff0000' };
    const { getByText } = render(
      <Popup destroyOnClose placement="top" overlayStyle={testStyle} overlayClassName={testClassName}>
        {triggerElement}
      </Popup>,
    );

    // 模拟鼠标进入
    act(() => {
      fireEvent.mouseEnter(getByText(triggerElement));
      jest.runAllTimers();
    });

    // 鼠标进入后，有元素，而且内容为空
    const popupContainer = await waitFor(() => document.querySelector(`.${testClassName}`));
    expect(popupContainer).not.toBeNull();
    expect(popupContainer).toHaveTextContent('');
    expect(popupContainer).toHaveStyle(testStyle);
  });

  test('测试浮层嵌套', async () => {
    const wrappedTriggerElement = '嵌套触发元素';
    const wrappedPopupTestId = 'wrapped-popup-test-id';
    const wrappedPopupText = '嵌套弹出层内容';
    const { getByText, queryByTestId } = render(
      <Popup
        placement="top"
        trigger="click"
        destroyOnClose
        content={
          <Popup
            placement="top"
            trigger="click"
            destroyOnClose
            content={<div data-testid={wrappedPopupTestId}>{wrappedPopupText}</div>}
          >
            <div data-testid={popupTestId}>{wrappedTriggerElement}</div>
          </Popup>
        }
      >
        {triggerElement}
      </Popup>,
    );

    // 初始时，所有浮层都不存在
    const popupElement1 = await waitFor(() => queryByTestId(popupTestId));
    const wrappedPopupElement1 = await waitFor(() => queryByTestId(wrappedPopupTestId));
    expect(popupElement1).toBeNull();
    expect(wrappedPopupElement1).toBeNull();

    // 触发浮层和嵌套浮层
    act(() => {
      fireEvent.click(getByText(triggerElement));
      jest.runAllTimers();
      fireEvent.click(getByText(wrappedTriggerElement));
      jest.runAllTimers();
    });

    // 所有浮层都展示出来
    const popupElement2 = await waitFor(() => queryByTestId(popupTestId));
    const wrappedPopupElement2 = await waitFor(() => queryByTestId(wrappedPopupTestId));
    expect(popupElement2).not.toBeNull();
    expect(wrappedPopupElement2).not.toBeNull();

    // 嵌套元素的浮层触发 mouseDown，不应该关闭任何浮层
    act(() => {
      fireEvent.mouseDown(queryByTestId(wrappedPopupTestId));
      jest.runAllTimers();
    });

    // 所有浮层都展示出来
    const popupElement3 = await waitFor(() => queryByTestId(popupTestId));
    const wrappedPopupElement3 = await waitFor(() => queryByTestId(wrappedPopupTestId));
    expect(popupElement3).not.toBeNull();
    expect(wrappedPopupElement3).not.toBeNull();
  });

  test('异常情况：浮层隐藏时点击其他地方，浮层不可以展示出来', async () => {
    const testClassName = 'test-class-name';
    render(
      <Popup placement="top" overlayClassName={testClassName}>
        {triggerElement}
      </Popup>,
    );

    // 浮层隐藏时，随便点击
    act(() => {
      fireEvent.click(document.body);
      jest.runAllTimers();
    });

    // 鼠标进入后，有元素，而且内容为空
    const popupContainer = await waitFor(() => document.querySelector(`.${testClassName}`));
    expect(popupContainer).toBeNull();
  });

  test('异常情况：getPopupContainer 传了非 HTMLElement 的元素', async () => {
    const testClassName = 'test-class-name';
    const { getByText } = render(
      <Popup visible placement="top" overlayClassName={testClassName} attach={() => 'xxx' as any}>
        {triggerElement}
      </Popup>,
    );

    // 浮层隐藏时，随便点击
    act(() => {
      fireEvent.click(getByText(triggerElement));
      jest.runAllTimers();
    });

    // 有元素，并且是渲染在 body 上
    const popupContainer = await waitFor(() => document.querySelector(`.${testClassName}`));
    expect(popupContainer.parentElement.parentElement.parentElement).toBe(document.body);
  });
});
