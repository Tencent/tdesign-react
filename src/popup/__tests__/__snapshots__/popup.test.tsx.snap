// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base.jsx 1`] = `
<DocumentFragment>
  <div
    class="t-popup__reference"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        Hover me
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`container.jsx 1`] = `
<DocumentFragment>
  <div
    class="tdesign-demo-block-row"
  >
    <div
      class="t-popup__reference"
    >
      <button
        class="t-button t-button--theme-primary t-button--variant-base"
        type="button"
      >
        <span
          class="t-button__text"
        >
          点击查看我的父元素(CSSSelector)
        </span>
      </button>
    </div>
    <div
      class="t-popup__reference"
    >
      <button
        class="t-button t-button--theme-primary t-button--variant-base"
        type="button"
      >
        <span
          class="t-button__text"
        >
          点击查看我的父元素(Function)
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`destroy.jsx 1`] = `
<DocumentFragment>
  <div
    class="t-popup__reference"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        Hover me
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`disabled.jsx 1`] = `
<DocumentFragment>
  <div
    class="t-popup__reference"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        Hover me
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`dynamic.jsx 1`] = `
<DocumentFragment>
  <div
    class="t-popup__reference"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        点击改变内容
      </span>
    </button>
  </div>
</DocumentFragment>
`;

exports[`placement.jsx 1`] = `
<DocumentFragment>
  <div
    style="margin: 0px auto; width: 500px; height: 260px; position: relative;"
  >
    <div
      class="t-popup__reference"
    >
      <button
        class="t-button t-button--theme-primary t-button--variant-base"
        style="position: absolute; top: 0px; left: 42%;"
        type="button"
      >
        <span
          class="t-button__text"
        >
          top
        </span>
      </button>
    </div>
    <div
      class="t-popup__reference"
    >
      <button
        class="t-button t-button--theme-primary t-button--variant-base"
        style="position: absolute; top: 0px; left: 70px;"
        type="button"
      >
        <span
          class="t-button__text"
        >
          top-left
        </span>
      </button>
    </div>
    <div
      class="t-popup__reference"
    >
      <button
        class="t-button t-button--theme-primary t-button--variant-base"
        style="position: absolute; top: 0px; right: 70px;"
        type="button"
      >
        <span
          class="t-button__text"
        >
          top-right
        </span>
      </button>
    </div>
    <div
      class="t-popup__reference"
    >
      <button
        class="t-button t-button--theme-primary t-button--variant-base"
        style="position: absolute; bottom: 0px; left: 42%;"
        type="button"
      >
        <span
          class="t-button__text"
        >
          bottom
        </span>
      </button>
    </div>
    <div
      class="t-popup__reference"
    >
      <button
        class="t-button t-button--theme-primary t-button--variant-base"
        style="position: absolute; bottom: 0px; left: 70px; width: 120px;"
        type="button"
      >
        <span
          class="t-button__text"
        >
          bottom-left
        </span>
      </button>
    </div>
    <div
      class="t-popup__reference"
    >
      <button
        class="t-button t-button--theme-primary t-button--variant-base"
        style="position: absolute; bottom: 0px; right: 70px;"
        type="button"
      >
        <span
          class="t-button__text"
        >
          bottom-right
        </span>
      </button>
    </div>
    <div
      class="t-popup__reference"
    >
      <button
        class="t-button t-button--theme-primary t-button--variant-base"
        style="position: absolute; left: 0px; top: 42%;"
        type="button"
      >
        <span
          class="t-button__text"
        >
          left
        </span>
      </button>
    </div>
    <div
      class="t-popup__reference"
    >
      <button
        class="t-button t-button--theme-primary t-button--variant-base"
        style="position: absolute; left: 0px; top: 50px;"
        type="button"
      >
        <span
          class="t-button__text"
        >
          left-top
        </span>
      </button>
    </div>
    <div
      class="t-popup__reference"
    >
      <button
        class="t-button t-button--theme-primary t-button--variant-base"
        style="position: absolute; left: 0px; bottom: 50px;"
        type="button"
      >
        <span
          class="t-button__text"
        >
          left-bottom
        </span>
      </button>
    </div>
    <div
      class="t-popup__reference"
    >
      <button
        class="t-button t-button--theme-primary t-button--variant-base"
        style="position: absolute; right: 0px; top: 42%;"
        type="button"
      >
        <span
          class="t-button__text"
        >
          right
        </span>
      </button>
    </div>
    <div
      class="t-popup__reference"
    >
      <button
        class="t-button t-button--theme-primary t-button--variant-base"
        style="position: absolute; right: 0px; top: 50px;"
        type="button"
      >
        <span
          class="t-button__text"
        >
          right-top
        </span>
      </button>
    </div>
    <div
      class="t-popup__reference"
    >
      <button
        class="t-button t-button--theme-primary t-button--variant-base"
        style="position: absolute; right: 0px; bottom: 50px;"
        type="button"
      >
        <span
          class="t-button__text"
        >
          right-bottom
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`style.jsx 1`] = `
<DocumentFragment>
  <div
    class="tdesign-demo-block-row"
  >
    <div
      class="t-popup__reference"
    >
      <button
        class="t-button t-button--theme-primary t-button--variant-base"
        type="button"
      >
        <span
          class="t-button__text"
        >
          click我试试
        </span>
      </button>
    </div>
    <div
      class="t-popup__reference"
    >
      <button
        class="t-button t-button--theme-primary t-button--variant-base"
        type="button"
      >
        <span
          class="t-button__text"
        >
          根据 trigger 元素定制 overlayStyle
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`trigger.jsx 1`] = `
<DocumentFragment>
  <div
    class="tdesign-demo-block-row"
  >
    <div
      class="t-popup__reference"
    >
      <button
        class="t-button t-button--theme-default t-button--variant-outline"
        type="button"
      >
        <span
          class="t-button__text"
        >
          悬浮时触发（默认）
        </span>
      </button>
    </div>
    <div
      class="t-popup__reference"
    >
      <div
        class="t-input__wrap"
        style="width: 200px;"
      >
        <div
          class="t-input t-size-m t-align-left t-is-default"
        >
          <input
            class="t-input__inner"
            placeholder="获得焦点时触发"
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="t-popup__reference"
    >
      <button
        class="t-button t-button--theme-default t-button--variant-outline"
        type="button"
      >
        <span
          class="t-button__text"
        >
          点击时触发
        </span>
      </button>
    </div>
    <div
      class="t-popup__reference"
    >
      <button
        class="t-button t-button--theme-default t-button--variant-outline"
        type="button"
      >
        <span
          class="t-button__text"
        >
          右击时触发
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`trigger-element.jsx 1`] = `
<DocumentFragment>
  <div
    class="tdesign-demo-block-row"
  >
    <div
      class="t-popup__reference"
    >
      <button
        class="t-button t-button--theme-default t-button--variant-outline"
        type="button"
      >
        <span
          class="t-button__text"
        >
          默认子节点元素触发
        </span>
      </button>
    </div>
    <div
      class="t-popup__reference"
    >
      <button
        class="t-button t-button--theme-primary t-button--variant-base"
        type="button"
      >
        <span
          class="t-button__text"
        >
          使用 triggerElement 元素触发
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`visible.jsx 1`] = `
<DocumentFragment>
  <div
    class="t-popup__reference"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        一直显示
      </span>
    </button>
  </div>
</DocumentFragment>
`;
